core/messaging/message.py

```diff
@@ -0,0 +1,356 @@
+from abc import ABC, abstractmethod
+
+import pandas as pd
+
+from core.const import (
+    INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION,
+    INTERNAL_TABLE_TO_FORM_SHEET,
+    INTERNAL_TYPE_TO_MESSAGE_FORMAT,
+    TABLE_AND_COLUMN_TO_ORIGINAL_COLUMN_LETTER,
+)
+from core.messaging import TF_message_bank as msgs
+from core.util import get_project_number_by_position, join_as_string
+from core.validation.exceptions import UnimplementedErrorMessageException
+from core.validation.failures.user import PreTransFormationFailure
+from core.validation.utils import get_cell_indexes_for_outcomes
+
+
+def remove_errors_already_caught_by_null_failure(
+    errors: list[tuple[str, str, str, str, str]]
+) -> list[tuple[str, str, str, str, str]]:
+    """
+    Removes errors from the list that have already been caught by null failures based on their sheet, section, and
+    cell index, and keeps only those present in null_failures. Additionally, includes all null_failures and errors
+    not present in null_failures or any part of the cell index is not already captured by a null failure.
+
+    :param errors: List of error tuples (sheet, section, cell_index, message, error_type).
+    :return: Filtered list of errors, including all null_failures and errors not present in null_failures or any part
+    of the cell index is not already captured by a null failure.
+    """
+
+    def normalise_risk_section_name(section):
+        if "Risks" in section:
+            return "Project / Programme Risks"
+        return section
+
+    null_messages = [msgs.BLANK, msgs.BLANK_ZERO, msgs.BLANK_PSI, msgs.BLANK_UNIT_OF_MEASUREMENT]
+
+    null_failures = [
+        (sheet, section, cell_index.strip(), message, error_type)
+        for sheet, section, cell_index, message, error_type in errors
+        if message in null_messages
+    ]
+
+    unique_identifiers_from_null_failures = {
+        (sheet, normalise_risk_section_name(section), cell_index.strip())
+        for sheet, section, cell_index, _, _ in null_failures
+        for cell_index in cell_index.split(",")
+    }
+
+    filtered_errors = [
+        error
+        for error in errors
+        if not any(
+            ((sheet, normalise_risk_section_name(section), cell_index.strip()) in unique_identifiers_from_null_failures)
+            for sheet, section, cell_index, _, _ in [error]
+            for cell_index in error[2].split(",")
+        )
+    ]
+
+    filtered_errors.extend(null_failures)
+    return filtered_errors
```

Comment:
Not a comment for this PR necessarily but this function is a monster.. we should look to improve the way we do this

Reply:
This is the final outstanding issue with this PR. I'd like not to have to refactor this twice so I need to do it right.

One option, to remove this function entirely is: we do not pick up composite key errors if the key contains a null with a line like:
`duplicated_rows = duplicated_rows.dropna()`

This mostly does the trick, however we run into the issue that Risks section composite key contains at least one null, proggramme/project ID. meaning we won't pick up any composite key errors for Risks